apps/mobile/src/hooks/useDelegateKey.ts

```diff
@@ -0,0 +1,176 @@
+import { useState, useCallback } from 'react'
+import DeviceInfo from 'react-native-device-info'
+import 'react-native-get-random-values'
+import { HDNodeWallet, Wallet } from 'ethers'
+import { useAuthGetNonceV1Query, useAuthVerifyV1Mutation } from '@safe-global/store/gateway/AUTO_GENERATED/auth'
+import {
+  useNotificationsDeleteSubscriptionV2Mutation,
+  useNotificationsUpsertSubscriptionsV2Mutation,
+} from '@safe-global/store/gateway/AUTO_GENERATED/notifications'
+
+import Logger from '@/src/utils/logger'
+import { Address } from '@/src/types/address'
+import { selectActiveSafe } from '@/src/store/activeSafeSlice'
+import { useAppSelector } from '@/src/store/hooks'
+import { useSiwe } from './useSiwe'
+import { useSign } from './useSign'
+import { isAndroid } from '../config/constants'
+import { selectFCMToken } from '../store/notificationsSlice'
+
+const ERROR_MSG = 'useDelegateKey: Something went wrong'
+
+export function useDelegateKey() {
+  // Local states
+  const [loading, setLoading] = useState<boolean>(false)
+  const [error, setError] = useState<unknown>(null)
+
+  // Queries
+  const [authVerifyV1] = useAuthVerifyV1Mutation()
+  const [notificationsUpsertSubscriptionsV2] = useNotificationsUpsertSubscriptionsV2Mutation()
+  const [notificationsDeleteSubscriptionsV2] = useNotificationsDeleteSubscriptionV2Mutation()
+  // Custom hooks
+  const { signMessage } = useSiwe()
+  const { getPrivateKey } = useSign()
+
+  // Redux states
+  const activeSafe = useAppSelector(selectActiveSafe)
+  const fcmToken = useAppSelector(selectFCMToken)
+
+  // Step 0 - Get the nonce to be included in the message to be sent to the backend
+  const { data } = useAuthGetNonceV1Query()
+
+  const createDelegate = useCallback(
+    async (ownerAddress: Address) => {
+      setLoading(true)
+      setError(null)
+      const nonce = data?.nonce
+      // Step 1 - Try to get the owner's private key
+      const ownerPrivateKey = await getPrivateKey()
+
+      try {
+        if (!ownerAddress || !activeSafe || !nonce || !fcmToken) {
```

Comment:
could be moved to a validate function.

Reply:
I think that if the try block is structured as:
```
        validateParameters(...)

        const ownerPrivateKey = await getPrivateKey(...)
        const signerAccount = await createSignerAccount(ownerPrivateKey)
        if (!signerAccount) throw new Error(ERROR_MSG)

        const siweMessage = generateSiweMessage(...)

        // Verify the delegate on the backend and subscribe for notifications.
        await verifyAndSubscribeDelegate(...)

```
it would make it way easier to read and follow

Reply:
Also what you get is separation of conserns. We abstract everything in small, testable functions. It not only improves readibility, but also makes it easier to unit test.